torch/autograd/gradcheck.py

```diff
@@ -299,12 +302,17 @@ def fn(input):
         def fn(input):
             return _as_tuple(func(*input))[i]
 
-        analytical, reentrant, correct_grad_sizes = get_analytical_jacobian(tupled_inputs, o, nondet_tol=nondet_tol)
+        analytical, reentrant, correct_grad_sizes, correct_grad_types = get_analytical_jacobian(tupled_inputs,
+                                                                                                o,
+                                                                                                nondet_tol=nondet_tol)
         numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
 
         if not correct_grad_sizes:
             return fail_test('Analytical gradient has incorrect size')
 
+        if not correct_grad_types:
```

Comment:
This is a dangerous path to go down to.
We do have some functions that return gradients with a different dtype. Actually most function work if you backprop float32 when you actually forwarded with a float16.

I am surprised that nothing trips this (it might on the amp repo though!)
I would at least add a flag to disable it if needed. And maybe disable it by default for BC reasons?

Reply:
This was the dtype check we saw previously and decided to remove. I can just revert it.

Reply:
Ho right! I guess that's why it's passing our CI!
I guess it's ok for now.
Maybe we want to do add a flag to disable this if the user wants? To have a more gentle BC-breaking story for this!

Reply:
Sure. I can add a default kwarg to gradcheck so existing users aren't affected.